fuel-vm/src/interpreter/metadata.rs

```diff
@@ -108,54 +108,29 @@ pub(crate) fn metadata(
     chain_id: ChainId,
     tx_offset: Word,
 ) -> SimpleResult<()> {
-    let external = context.is_external();
-    let args = GMArgs::try_from(imm)?;
+    let parent = context
+        .is_internal()
+        .then(|| frames.last().map(|f| f.registers()[RegId::FP]))
+        .flatten();
 
-    if external {
-        match args {
-            GMArgs::GetVerifyingPredicate => {
-                *result = context
-                    .predicate()
-                    .map(|p| p.idx() as Word)
-                    .ok_or(PanicReason::TransactionValidity)?;
-            }
-
-            GMArgs::GetChainId => {
-                *result = chain_id.into();
-            }
-
-            GMArgs::TxStart => {
-                *result = tx_offset;
-            }
-
-            _ => return Err(PanicReason::ExpectedInternalContext.into()),
-        }
-    } else {
-        let parent = frames
-            .last()
-            .map(|f| f.registers()[RegId::FP])
-            .expect("External context will always have a frame");
-
-        match args {
-            GMArgs::IsCallerExternal => {
-                *result = (parent == 0) as Word;
-            }
-
-            GMArgs::GetCaller if parent != 0 => {
-                *result = parent;
-            }
-
-            GMArgs::GetChainId => {
-                *result = chain_id.into();
-            }
-
-            GMArgs::TxStart => {
-                *result = tx_offset;
-            }
-
-            _ => return Err(PanicReason::ExpectedInternalContext.into()),
-        }
-    }
+    *result = match GMArgs::try_from(imm)? {
+        GMArgs::GetVerifyingPredicate => context
```

Comment:
Before we were returning an `ExpectedInternalContext` error in the case of the internal context. Do we need to do the same?

Reply:
Good catch. The new behavior is what the specs say. I see no reason to disallow internal contexts here.

Reply:
The specification says that it is allowed only in the context of the predicate:

<img width="731" alt="image" src="https://github.com/FuelLabs/fuel-vm/assets/18346821/37673c72-c61e-4ac5-942d-2137dcd43c7c">

And in our codebase predicate context is external context:

<img width="676" alt="image" src="https://github.com/FuelLabs/fuel-vm/assets/18346821/a21a386c-2b0f-4e2b-a17f-1f55a431237f">

So it looks like we need ti check that it is external context=)

Reply:
But the code already performs the check: The `context.predicate()` returns `None` outside predicate contexts, leading to `TransactionValidity` panic. Previously calling `gm/GetVerifyingPredicate` panicked with `ExpectedInternalContext` which was definitely incorrect. Now if we check with something like
```rust
if context.is_external() { // pseudocode
  result = context.predicate().map(...).ok_or(PanicReason::TransactionValidity)?;
} else {
  return PanicReason::ExpectedExternalContext;
}
```
then the only additional check we're actually performing is that `Call` contexts return `ExpectedExternalContext` instead of `TransactionValidity`. Is the issue just returning a wrong panic reason? Am I missing something here?

Reply:
Ohh, I missed the moment that `predicate` already is doing this check inside=D Hmm, maybe using of another error will make more sense, because `TransactionValidity` is a strange choice

Reply:
I'm going to merge this as-is, since `TransactionValidity` is (only) used for this kind of things currently. We can do a rename in a separate PR if needed.